R/ppc-loo.R

```diff
@@ -456,3 +490,81 @@ ppc_loo_ribbon <-
   return(psis_object)
 }
 
+## Boundary correction based on code by Yang Hu and Carl Scarrott in evmix package
+
+# Boundary correction KDE helper function
+.bc_dunif <- function(xs, pvals, b, xmax = 1){
+  # Function based on biased-corrected (modified) beta kernel 
+  # Chen, Song Xi. "Beta kernel estimators for density functions." 
+  # Computational Statistics & Data Analysis 31.2 (1999): 131-145.
+  
+  # Re-scaling inputs based on upper_bound given beta kernel is defined [0,1]
+  xs <- xs/xmax
+  pvals <- pvals/xmax
+  b <-  b/xmax # smoothing parameter (i.e. bw)
+  
+  d <- vapply(X = xs, 
+              FUN = .bc_kde_calc, 
+              b = b,
+              pvals = pvals, 
+              FUN.VALUE = 0)
+  
+  return(d/xmax)
+  
+}
+
+# Bias correction function (Chen 1999)
+.rho <- function(x, b) {
+  return(2*b^2 + 2.5 - sqrt(4*b^4 + 6*b^2 + 2.25 - x^2 - x/b))
+}
+
+# Piecewise kernel value estimation (Chen 1999)
+.bc_kde_calc <- function(xs, b,  pvals){
+
+  if ((xs >= 2*b) & (xs <= (1 - 2*b))) {
+    d <- mean(dbeta(pvals, xs/b, (1 - xs)/b))
+  } else if ((xs >= 0) & (xs < 2*b)) {
+    d <- mean(dbeta(pvals, .rho(xs, b), (1 - xs)/b))
+  } else if ((xs > (1 - 2*b)) & (xs <= 1)) {
+    d <- mean(dbeta(pvals, xs/b, .rho(1 - xs, b)))
+  } else {
+    d <- 0
+  }
+  return(d)
+}
+
+# Wrapper function, transforms pvals / PIT vals into bc values
+.bc_pvals <- function(x, bw = "nrd0"){
+  
+  # Set-up
+  xs  <- seq(0, 1, length.out = length(x))
+  d <- xs
+  xmax <- 1 + 1e-9
+  bw <- density(x, bw = bw)$bw # extract bw
+  
+  # get only valid pvals
+  valid_pvals <- x[is.finite(x)]
+  
+  # Some sanity checks
+  if (abs(xmax - max(valid_pvals)) >= 1e-1) {
+    stop("largest PIT value must be below 1")
+  }
+  
+  # Ignore zeros to avoid problems during KDE estimation
+  if (any(valid_pvals == 0)) {
+    warning(paste("Ignored", sum(valid_pvals == 0),
+                  "PIT values == 0, they are invalid for beta KDE method"))
+    valid_pvals = valid_pvals[valid_pvals != 0]
+  }
+  
+  # Boundary corrected KDE values
+  bc_vals <- .bc_dunif(xs = xs, pvals=valid_pvals, b =bw)
+  
+  # Set any negative values to zero and output bc density values
+  bc_vals[which(bc_vals < 0)] = 0
+  d[ifelse(!is.na(xs), (xs >= 0) & (xs <= 1), FALSE)] = bc_vals
```

Comment:
In theory yes, we take care of length consistency by taking care of PIT values near the 0 and 1 bounds via `bc_pvals()` in the  following section

```r
## I will change this based on your comment below
if (abs(xmax - max(valid_pvals)) >= 1e-1) {
    stop("largest PIT value must be below 1")

if (any(valid_pvals == 0)) {
    warning(paste("Ignored", sum(valid_pvals == 0),
                  "PIT values == 0, they are invalid for beta KDE method"))
    valid_pvals = valid_pvals[valid_pvals != 0]
  }
```

This basically defines the length of the subsequent reference uniforms which are generated via runif [0,1] from R, so this line of code might be redundant but I thought to add it just in case.